File: src/quant_ai/analytics/benchmark_attribution.py
"""Single-period Brinson-Fachler arithmetic attribution from reviewed inputs.

Beginning weights and same-period total returns, one currency, long-only and
fully invested including cash. No inferred sector returns or multi-period linking.
"""
from datetime import date
from decimal import Decimal, InvalidOperation, localcontext


def benchmark_attribution(document: dict) -> dict:
    def require(condition, message):
        if not condition:
            raise ValueError(message)

    def number(value):
        require(isinstance(value, str) and len(value) <= 80, "decimal_strings_required")
        try:
            result = Decimal(value)
        except InvalidOperation as error:
            raise ValueError("invalid_decimal") from error
        require(result.is_finite() and abs(result) <= 1000, "invalid_decimal")
        require(result.as_tuple().exponent >= -18, "excess_decimal_precision")
        return result

    require(isinstance(document, dict), "invalid_document")
    require(document.get("schema") == "pramana.benchmark_attribution_input.v1", "invalid_schema")
    require(document.get("basis") == "beginning_weights_same_period_total_returns", "unsupported_basis")
    require(document.get("currency") == "INR", "unsupported_currency")
    identity = {}
    for key in ("portfolioId", "benchmarkId"):
        value = document.get(key)
        require(isinstance(value, str) and value == value.strip() and 0 < len(value) <= 100
                and all(character.isprintable() for character in value), "invalid_" + key)
        identity[key] = value
    dates = []
    for key in ("periodStart", "periodEnd"):
        value = document.get(key)
        require(isinstance(value, str) and len(value) == 10, "invalid_" + key)
        try:
            parsed_date = date.fromisoformat(value)
        except ValueError as error:
            raise ValueError("invalid_" + key) from error
        require(parsed_date.isoformat() == value, "invalid_" + key)
        dates.append(parsed_date)
        identity[key] = value
    require(dates[0] < dates[1], "invalid_period_order")
    rows = document.get("sectors")
    require(isinstance(rows, list) and 0 < len(rows) <= 500, "invalid_sectors")
    with localcontext() as context:
        context.prec = 80
        parsed, names = [], set()
        for row in rows:
            require(isinstance(row, dict), "invalid_sector")
            name = row.get("name")
            require(isinstance(name, str) and name == name.strip() and 0 < len(name) <= 100 and name not in names, "invalid_or_duplicate_sector")
            names.add(name)
            wp, wb, rp, rb = [number(row.get(key)) for key in
                              ("portfolioWeight", "benchmarkWeight", "portfolioReturn", "benchmarkReturn")]
            require(0 <= wp <= 1 and 0 <= wb <= 1, "invalid_weight")
            require(rp >= -1 and rb >= -1, "invalid_long_only_return")
            parsed.append((name, wp, wb, rp, rb))
        require(sum(r[1] for r in parsed) == 1 and sum(r[2] for r in parsed) == 1, "weights_must_sum_to_one_include_cash")
        portfolio_return = sum(wp * rp for _, wp, _, rp, _ in parsed)
        benchmark_return = sum(wb * rb for _, _, wb, _, rb in parsed)
        require(portfolio_return == number(document.get("portfolioReturn")), "portfolio_return_not_reconciled")
        require(benchmark_return == number(document.get("benchmarkReturn")), "benchmark_return_not_reconciled")
        effects = []
        for name, wp, wb, rp, rb in parsed:
            allocation = (wp - wb) * (rb - benchmark_return)
            selection = wb * (rp - rb)
            interaction = (wp - wb) * (rp - rb)
            effects.append({"name": name, "allocation": allocation, "selection": selection,
                            "interaction": interaction, "total": allocation + selection + interaction})
        totals = {key: sum(row[key] for row in effects) for key in ("allocation", "selection", "interaction", "total")}
        active = portfolio_return - benchmark_return
        require(totals["total"] == active, "attribution_not_reconciled")
        return {"schema": "pramana.benchmark_attribution.v1", "method": "single_period_brinson_fachler",
                **identity, "sourceQualified": False, "currency": "INR", "portfolioReturn": str(portfolio_return), "benchmarkReturn": str(benchmark_return),
                "activeReturn": str(active), "reconciliationDifference": str(totals["total"] - active),
                "sectors": [{key: str(value) if isinstance(value, Decimal) else value for key, value in row.items()} for row in effects],
                "totals": {key: str(value) for key, value in totals.items()},
                "scope": "Arithmetic verification of supplied inputs; source qualification and dashboard publication remain separate."}

File: apps/pramana-ui/lib/benchmark-attribution.ts
import {createHash} from "node:crypto";

export type BenchmarkAttribution = {
  schema: "pramana.benchmark_attribution.v1"; method: "single_period_brinson_fachler";
  portfolioId: string; benchmarkId: string; periodStart: string; periodEnd: string;
  currency: "INR"; sourceQualified: false; portfolioReturn: string; benchmarkReturn: string;
  activeReturn: string; reconciliationDifference: string;
  sectors: {name: string; allocation: string; selection: string; interaction: string; total: string}[];
  totals: {allocation: string; selection: string; interaction: string; total: string};
  inputSha256: string; inputPayload: string; input: unknown; scope: string;
};
function check(value: unknown): asserts value {if (!value) throw Error("invalid_benchmark_attribution");}
function decimal(value: unknown): number {
  check(typeof value === "string" && value.length <= 80 && /^[+-]?(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?$/.test(value));
  const n = Number(value); check(Number.isFinite(n) && Math.abs(n) <= 1000000); return n;
}
function same(value: unknown, expected: number) {check(Math.abs(decimal(value)-expected) <= 1e-12*Math.max(1,Math.abs(expected)));}
function label(value: unknown): asserts value is string {check(typeof value === "string" && value.trim() === value && value.length > 0 && value.length <= 100 && !/[\x00-\x1f\x7f]/.test(value));}
function day(value: unknown): asserts value is string {
  check(typeof value === "string" && /^\d{4}-\d{2}-\d{2}$/.test(value));
  const time = Date.parse(value); check(Number.isFinite(time) && new Date(time).toISOString().slice(0,10) === value);
}

/** Independent arithmetic check with 1e-12 floating tolerance; not source certification. */
export function parseBenchmarkAttribution(raw: string, portfolioId: string): BenchmarkAttribution {
  check(Buffer.byteLength(raw) <= 4_000_000);
  const r = JSON.parse(raw);
  check(r && r.schema === "pramana.benchmark_attribution.v1" && r.method === "single_period_brinson_fachler" && r.currency === "INR" && r.sourceQualified === false);
  check(typeof r.inputPayload === "string" && Buffer.byteLength(r.inputPayload) <= 1_000_000);
  check(createHash("sha256").update(r.inputPayload).digest("hex") === r.inputSha256);
  const input = JSON.parse(r.inputPayload);
  check(JSON.stringify(input) === JSON.stringify(r.input));
  check(input.schema === "pramana.benchmark_attribution_input.v1" && input.basis === "beginning_weights_same_period_total_returns" && input.currency === "INR");
  for (const key of ["portfolioId","benchmarkId"] as const) {label(r[key]); check(input[key] === r[key]);}
  check(r.portfolioId === portfolioId);
  day(r.periodStart); day(r.periodEnd); check(r.periodStart < r.periodEnd);
  check(input.periodStart === r.periodStart && input.periodEnd === r.periodEnd);
  check(Array.isArray(input.sectors) && input.sectors.length > 0 && input.sectors.length <= 500);
  const names = new Set<string>();
  const rows = input.sectors.map((s: Record<string, unknown>) => {
    check(s && typeof s === "object"); label(s.name); check(!names.has(s.name)); names.add(s.name);
    const wp=decimal(s.portfolioWeight), wb=decimal(s.benchmarkWeight), rp=decimal(s.portfolioReturn), rb=decimal(s.benchmarkReturn);
    check(wp>=0 && wp<=1 && wb>=0 && wb<=1 && rp>=-1 && rb>=-1 && rp<=1000 && rb<=1000);
    return {name:s.name,wp,wb,rp,rb};
  }) as {name:string;wp:number;wb:number;rp:number;rb:number}[];
  same(String(rows.reduce((n,s)=>n+s.wp,0)),1); same(String(rows.reduce((n,s)=>n+s.wb,0)),1);
  const pr=rows.reduce((n,s)=>n+s.wp*s.rp,0), br=rows.reduce((n,s)=>n+s.wb*s.rb,0);
  same(input.portfolioReturn,pr); same(input.benchmarkReturn,br); same(r.portfolioReturn,pr); same(r.benchmarkReturn,br);
  check(Array.isArray(r.sectors) && r.sectors.length === rows.length);
  const totals={allocation:0,selection:0,interaction:0,total:0};
  rows.forEach((s,i)=>{
    const allocation=(s.wp-s.wb)*(s.rb-br), selection=s.wb*(s.rp-s.rb), interaction=(s.wp-s.wb)*(s.rp-s.rb);
    const values={allocation,selection,interaction,total:allocation+selection+interaction};
    check(r.sectors[i]?.name===s.name);
    for(const key of Object.keys(values) as (keyof typeof values)[]) {same(r.sectors[i][key],values[key]);totals[key]+=values[key];}
  });
  for(const key of Object.keys(totals) as (keyof typeof totals)[]) same(r.totals?.[key],totals[key]);
  same(r.activeReturn,pr-br); same(r.reconciliationDifference,0); same(String(totals.total),pr-br);
  check(typeof r.scope === "string" && r.scope.length <= 1000);
  return r as BenchmarkAttribution;
}

export type BenchmarkAttributionSummary = Omit<BenchmarkAttribution,"input"|"inputPayload"|"scope">;
export type BenchmarkAttributionState = {status:"unavailable"|"invalid"|"published";detail:string;report:BenchmarkAttributionSummary|null};
